cli/bench_command.go

```diff
@@ -411,72 +622,841 @@ func (c *benchCmd) bench(_ *fisk.ParseContext) error {
 
 	bm.Close()
 
-	if !c.noProgress {
-		uiprogress.Stop()
+	err = c.printResults(bm)
+	if err != nil {
+		return err
 	}
 
-	if c.fetchTimeout {
-		log.Print("WARNING: at least one of the pull consumer Fetch operation timed out. These results are not optimal!")
-	}
+	return nil
+}
 
-	if c.retriesUsed {
-		log.Print("WARNING: at least one of the JS publish operations had to be retried. These results are not optimal!")
+func (c *benchCmd) serveAction(_ *fisk.ParseContext) error {
+	// reply mode is open-ended for the number of messages
+	err := c.processActionArgs()
+	if err != nil {
+		return err
 	}
 
-	fmt.Println()
-	fmt.Println(bm.Report())
+	banner := c.generateBanner(BenchTypeServiceServe)
 
-	if c.csvFile != "" {
-		csv := bm.CSV()
-		err := os.WriteFile(c.csvFile, []byte(csv), 0600)
+	log.Println(banner)
+
+	bm := bench.NewBenchmark("NATS", c.numClients, 0)
+
+	startwg := &sync.WaitGroup{}
+	donewg := &sync.WaitGroup{}
+
+	for i := 0; i < c.numClients; i++ {
+		nc, err := nats.Connect(opts().Config.ServerURL(), natsOpts()...)
 		if err != nil {
-			log.Printf("error writing file %s: %v", c.csvFile, err)
+			return fmt.Errorf("nats connection %d failed: %s", i, err)
 		}
-		fmt.Printf("Saved metric data in csv file %s\n", c.csvFile)
+		defer nc.Close()
+
+		startwg.Add(1)
+		donewg.Add(1)
+
+		go c.runServiceServer(nc, startwg, donewg)
+	}
+
+	startwg.Wait()
+	donewg.Wait()
+
+	bm.Close()
+
+	err = c.printResults(bm)
+	if err != nil {
+		return err
 	}
 
 	return nil
 }
 
-func min(a, b int) int {
-	if a < b {
-		return a
+func (c *benchCmd) jspubAction(_ *fisk.ParseContext) error {
+	err := c.processActionArgs()
+	if err != nil {
+		return err
 	}
-	return b
-}
 
-func getSubscribeSubject(c *benchCmd) string {
-	if c.multiSubject {
-		return c.subject + ".*"
-	} else {
-		return c.subject
+	banner := c.generateBanner(BenchTypeJSPub)
+	log.Println(banner)
+	bm := bench.NewBenchmark("NATS", 0, c.numClients)
+	benchId := strconv.FormatInt(time.Now().UnixMilli(), 16)
+	startwg := &sync.WaitGroup{}
+	donewg := &sync.WaitGroup{}
+
+	// create the stream for the benchmark (and purge it)
+	nc, err := nats.Connect(opts().Config.ServerURL(), natsOpts()...)
+	if err != nil {
+		log.Fatalf("NATS connection failed: %v", err)
```

Comment:
I'd prefer we dont use `log.Fatalf` but instead return errors so that the cli processor can handle the exit and printing of errors etc

cli can be used as a package so a package dying without giving callers a way to deal with that isnt super nice

with these being many go routines etc I can imagine getting this working is quite annoying but you could make a errors channel on `c` and check that at the call spots

Reply:
So basically re-did error handling in general everywhere for the bench command, which was indeed quite annoying and a fair bit of work, especially dealing with the go routines but now the code is completely free of log.Fatalf and returns the error all the way back up but the errors are better now (homogenized, wrapped, etc...) for it. Note that because each of the go routines could experience an error and only one can be returned it will print all of the errors for all the go routines but only return the first one back up.